core/src/main/java/io/parsingdata/metal/token/TokenRef.java

```diff
@@ -60,22 +64,22 @@ public TokenRef(final String name, final String referenceName, final Encoding en
 
     @Override
     protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
-        return lookup(environment.order, referenceName).parse(scope, environment, encoding);
+        return lookup(ImmutableList.create(environment.order), referenceName).computeResult().parse(scope, environment, encoding);
     }
 
-    private Token lookup(final ParseItem item, final String referenceName) {
-        if (item.getDefinition().name.equals(referenceName)) {
-            return item.getDefinition();
+    private SafeTrampoline<Token> lookup(final ImmutableList<ParseItem> items, final String referenceName) {
+        if (items.isEmpty()) { return complete(() -> LOOKUP_FAILED); }
+        final ParseItem item = items.head;
+        if (item.getDefinition().name.equals(referenceName)) { return complete(item::getDefinition); }
```

Comment:
And to add to that.

Previously 'simple' return statements have now become 'not so simple'. E.g.:

This was previously accepted by our own code style rules:

```java
if (something) { return head; }
```

In _many_ cases, this has now turned into:

```java
if (something) { return complete(() -> head); }
```

No so 'simple' anymore... Food for thought.